SCUMM: Check for and flag demos correctly (PR#931 Fixup)

digitall opened the pull request:
Rewrote the detection matching to use the extra field like the other
two.

This requires extensive testing due to touching code shared for a lot of
games and being sensitive to individual versions of games.

bluegr commented on engines/scumm/detection.cpp:
```diff
@@ -489,42 +489,45 @@ static void computeGameSettingsFromMD5(const Common::FSList &fslist, const GameF
+		// The gameid either matches, or is empty. The latter indicates
+		// a generic entry, currently used for some generic HE settings.
+
+		if (g->variant != 0 && !!scumm_stricmp(md5Entry->variant, g->variant)) continue;
```
The two not operators here (!!) are superfluous, and unneeded

bluegr commented on engines/scumm/detection.cpp:
```diff
@@ -489,42 +489,45 @@ static void computeGameSettingsFromMD5(const Common::FSList &fslist, const GameF
+		if (g->variant != 0 && !!scumm_stricmp(md5Entry->variant, g->variant)) continue;
+		// Variant matches too
+
+		if (g->preferredTag !=0 && !!scumm_stricmp(md5Entry->extra, g->preferredTag)) continue;
```
Same here, with !!

bluegr replied:
Also, this line is the only change to the detector code in this PR :)

bluegr commented on engines/scumm/detection.cpp:
```diff
@@ -489,42 +489,45 @@ static void computeGameSettingsFromMD5(const Common::FSList &fslist, const GameF
+		// Variant matches too
+
+		if (g->preferredTag !=0 && !!scumm_stricmp(md5Entry->extra, g->preferredTag)) continue;
+		//OH, and do check the extra field in the entry too, otherwise we miss demos
```
Nitpicking: this comment needs formatting to be the same as the rest

digitall commented:
@bluegr : I have addressed all your comments and rebased the commit in this PR onto the latest master commit.

One point, is that the "!!" is not just spurious, but seems to be a practice to avoid compiler warnings and sanitize the value to a boolean. See:
http://stupefydeveloper.blogspot.co.uk/2009/09/c-double-exclamation.html

I am not sure, but I think this is not necessary here unless @henke37 did it for some specific reason?

bluegr commented:
Thanks!

Yeah, the usage of !! does help sanitize a return value to a boolean, but it's not really needed here, as the condition above these does not use it.

Have you tried these changes? Are all the SCUMM games detected properly?

As mentioned, the only real change to the detector is the additional check for the preferredTag field

digitall commented:
@bluegr No problem. No, I haven't tried these changes yet, beyond checking they compile without warnings.

I can try moving my .scummvmrc file off, and doing a Mass Add with and without this patch to see what difference it makes, but I don't think I have most of the SCUMM game demos so this may be pointless. I suggest that a SCUMM engine dev who has all of the relevant demos tests this and sees if the detection is better as they are better placed to know if this is correct. Not sure who that should be? @sev- ?

sev- commented:
OK, merging this. Will do some detection testing after the merge.
